### PR TITLE
g.copy: Report failed copy using an error message

### PR DESCRIPTION
**Pull request by wenzeslaus**
Changes in behavior:

1. Non-zero return code (EXIT_FAILURE) has now also a correspoding
   error message (ERROR:) not just warning messages (from
   do_copy.c).
2. Skipping an input as ineligible for copying may now result in
   an error (fatal error with EXIT_FAILURE). The error newly occures when
   all inputs are skipped or, most importantly, when only one input
   was provided and it was skipped.

The skipping (in main.c) is now consistently reported as skipping.
The warning messages also contain more details and the new error
messages point to them.

The M_do_copy function (do_copy.c) now starts with 'Copying' because
it is at the begining of the process instead of 'Copy' which is misleading
(is the user supposed to copy).

The user manual has now a new intro and newly describes the behavior on
error discussing the much simpler single input case separatelly.

### CODE REVIEW

**Comment by wenzeslaus** [comment]
I have tried to make it more clear how g.copy behaves and make it work more as expected, i.e., reporting an error when error occurred (not just warnings combined with non-zero return code) and reporting an error when nothing was copied (when everything was skipped). The later being useful for scripting, esp. when only one item is copied at a time.

At the same time, I kept the current behavior of continuing with the other copy operations even when one failed. The rules to what happens are complicated when multiple items are involved. However, describing them was simpler than rewriting g.copy's main.c, M_do_copy() function from do_copy.c, and Vect_copy() from Vlib/map.c.

The most important change is that now when the input is skipped from g.copy's point of view, but it is the only input, an error is reported. The previous behavior is a warning and a *zero* return code, i.e. indicating no error. Now standard fatal error is used (i.e., message with ERROR: and non-zero return code). The error message revers to the "warning above" which is still the original warning.

File: general/g.copy/g.copy.html [hunk 1]
```diff
@@ -1,13 +1,48 @@
...
+
+<h3>Relation to mapsets</h3>
+
 A user may access data stored under the other mapsets listed in their
 mapset search path. However, the user may only modify data stored
 under their own current mapset. <em>g.copy</em> allows the user to copy
 existing data files <b>from</b> other mapsets <b>to</b> the user's
-current mapset ($MAPSET). The files to be copied must exist in the
+current mapset (<code>g.mapset -p</code>). The files to be copied must exist in the
 user's current mapset search path and location; output is sent to the
```
Comment (ninsbl) [inline general/g.copy/g.copy.html:23]: Maybe also add a hint on how to see/set the search path like for the current mapset
```suggestion
user's current mapset search path (<code>g.mapsets -s</code>) and location; output is sent to the
```

File: general/g.copy/g.copy.html [hunk 2]
```diff
@@ -1,13 +1,48 @@
...
 relevant data element directory(ies) under the user's current mapset.
 
+<h3>Behavior on error</h3>
+
+When only one map or other element is requested to be copied and the copying
+is not possible or fails, an error is reported.
+
+<p>
+Multiple maps or other elements can be copied using the same command.
+In that case, <em>g.copy</em> attempts to copy as much as possible
```
Comment (ninsbl) [inline general/g.copy/g.copy.html:33]:
```suggestion
If multiple maps or other elements are copied in one command,
<em>g.copy</em> attempts to copy as much as possible
```
It is already written above that multiple elements can be copied in one command, does not need to be repeated.

File: general/g.copy/g.copy.html [hunk 3]
```diff
@@ -1,13 +1,48 @@
...
 under their own current mapset. <em>g.copy</em> allows the user to copy
 existing data files <b>from</b> other mapsets <b>to</b> the user's
-current mapset ($MAPSET). The files to be copied must exist in the
+current mapset (<code>g.mapset -p</code>). The files to be copied must exist in the
 user's current mapset search path and location; output is sent to the
 relevant data element directory(ies) under the user's current mapset.
 
+<h3>Behavior on error</h3>
+
+When only one map or other element is requested to be copied and the copying
```
Comment (ninsbl) [inline general/g.copy/g.copy.html:28]:
```suggestion
Errors typically occur when a map or other element does not exist,
<b>from</b> and <b>to</b> are the same,  <b>to</b> element
already exists and overwriting (e.g., by <b>--overwrite</b>) is not
enabled or <b>to</b> element has an illegal name.
 
When only one map or other element is requested to be copied and the copying
```
Moved up from a paragraph further down.

File: general/g.copy/g.copy.html [hunk 4]
```diff
@@ -1,13 +1,48 @@
...
+If nothing can be copied or an error occurred during one of the copy
+operations, an error message is reported.
+Copying is skipped when a map or other element does not exist,
+<b>from</b> and <b>to</b> are the same, <b>to</b> element already exists
+and overwriting (e.g., by <b>--overwrite</b>) is not enabled
+or <b>to</b> element has an illegal name.
+Copying of remaining elements is not prevented by skipping or copying error
+of another element.
+However, as stated above, module ends with an error when copying failed,
+but only after other copy operations were performed.
```
Comment (ninsbl) [inline general/g.copy/g.copy.html:44]:
```suggestion
In that case, copying of the element causing problems is skipped, and
<em>g.copy</em> proceeds with copying the  remaining elements.

If nothing can be copied or an error occurred during one of the copy
operations, an error message is reported  after other possible copy
operations were performed.
```
Moved one sentence up, dedulicate language a little.

**Review by ninsbl — APPROVED** [review]
Looks good.
Some optional suggestions for the manual. Feel free to skip them if you prefer the original wording...

File: general/g.copy/g.copy.html [hunk 5]
```diff
@@ -1,13 +1,50 @@
...
-user's current mapset search path and location; output is sent to the
+current mapset (<code>g.mapset -p</code>). The files to be copied must exist in the
+user's current mapset search path (<code>g.mapsets -p</code>) and location;
+output is sent to the
 relevant data element directory(ies) under the user's current mapset.
 
+<h3>Behavior on error</h3>
+
+Errors typically occur when a map or other element does not exist,
+<b>from</b> and <b>to</b> are the same,  <b>to</b> element
```
Comment (wenzeslaus) [inline general/g.copy/g.copy.html:30]:
```suggestion
<b>from</b> and <b>to</b> are the same, <b>to</b> element
```

File: general/g.copy/g.copy.html [hunk 6]
```diff
@@ -1,13 +1,50 @@
...
+already exists and overwriting (e.g., by <b>--overwrite</b>) is not
+enabled or <b>to</b> element has an illegal name.
```
Comment (wenzeslaus) [inline general/g.copy/g.copy.html:32]:
```suggestion
enabled, or the <b>to</b> element has an illegal name.
```

This sentence is exactly the reason we have the Oxford comma.

File: general/g.copy/g.copy.html [hunk 7]
```diff
@@ -1,13 +1,50 @@
...
+ 
+When only one map or other element is requested to be copied and the copying
+is not possible or fails, an error is reported.
+
+<p>
+If multiple maps or other elements are copied in one command,
+<em>g.copy</em> attempts to copy as much as possible
+even when problems occur with one of the elements.
+In that case, copying of the element causing problems is skipped, and
+<em>g.copy</em> proceeds with copying the  remaining elements.
```
Comment (wenzeslaus) [inline general/g.copy/g.copy.html:42]:
```suggestion
<em>g.copy</em> proceeds with copying the remaining elements.
```

**Review by wenzeslaus — COMMENTED** [review]
@ninsbl I liked your text. Much better. I made some further changes in the revised text.

I included `g.mapsets -p` instead of -s you suggested. It is really the equivalent of `g.mapset -p`. The point is just to show what is current. Not necessarily teach how to modify the path and -s is an oddball anyway.

**Pull request merged**